app/src/main/java/org/schabi/newpipe/player/ui/PlayerUiList.kt

```diff
@@ -0,0 +1,124 @@
+                playerUi.initPlayback()
+            }
+        }
+
+        playerUis.runWithLockSync {
+            lockData.add(playerUi)
+        }
+    }
+
+    /**
+     * Destroys all matching player UIs and removes them from the list.
+     * @param playerUiType the class of the player UI to destroy;
+     * the [Class.isInstance] method will be used, so even subclasses will be
+     * destroyed and removed
+     * @param T the class type parameter </T>
+     * */
+    fun <T> destroyAll(playerUiType: Class<T?>) {
+        val toDestroy = mutableListOf<PlayerUi>()
+
+        // short blocking removal from class to prevent interfering from other threads
+        playerUis.runWithLockSync {
+            val new = mutableListOf<PlayerUi>()
+            for (ui in lockData) {
+                if (playerUiType.isInstance(ui)) {
+                    toDestroy.add(ui)
+                } else {
+                    new.add(ui)
+                }
+            }
+            lockData = new
+        }
+        // then actually destroy the UIs
+        for (ui in toDestroy) {
+            ui.destroyPlayer()
+            ui.destroy()
+        }
+    }
+
+    /**
+     * @param playerUiType the class of the player UI to return;
+     * the [Class.isInstance] method will be used, so even subclasses could be returned
+     * @param T the class type parameter
+     * @return the first player UI of the required type found in the list, or null
+     </T> */
+    fun <T> get(playerUiType: Class<T>): T? =
+        playerUis.runWithLockSync {
+            for (ui in lockData) {
+                if (playerUiType.isInstance(ui)) {
+                    when (val r = playerUiType.cast(ui)) {
+                        // try all UIs before returning null
+                        null -> continue
+                        else -> return@runWithLockSync r
+                    }
+                }
+            }
+            return@runWithLockSync null
+        }
+
+    /**
+     * @param playerUiType the class of the player UI to return;
+     * the [Class.isInstance] method will be used, so even subclasses could be returned
+     * @param T the class type parameter
+     * @return the first player UI of the required type found in the list, or an empty
+     * [         ] otherwise
+     </T> */
+    @Deprecated("use get", ReplaceWith("get(playerUiType)"))
+    fun <T> getOpt(playerUiType: Class<T>): Optional<T & Any> =
+        Optional.ofNullable(get(playerUiType))
+
+    /**
+     * Calls the provided consumer on all player UIs in the list, in order of addition.
+     * @param consumer the consumer to call with player UIs
+     */
+    fun call(consumer: java.util.function.Consumer<PlayerUi>) {
+        // copy the list out of the mutex before calling the consumer which might block
+        val new = playerUis.runWithLockSync {
+            lockData.toMutableList()
+        }
+        for (ui in new) {
+            consumer.accept(ui)
+        }
+    }
```

Comment:
```suggestion
    fun call(consumer: (PlayerUi) -> Unit) {
        // copy the list out of the mutex before calling the consumer which might block
        val new = playerUis.runWithLockSync {
            lockData.toMutableList()
        }
        for (ui in new) {
            consumer(ui)
        }
    }
```

Reply:
Does this work even when called from Java?

Reply:
Yeah, but Unit.INSTANCE has to be returned.

Reply:
I changed the parameter type back to `Consumer`; the synchronized list implementation automatically handles the synchronization needed.